web/utils.go

```diff
@@ -73,23 +73,61 @@ func GetCertFiles(certFile, keyFile string) (string, string) {
 	return certFile, keyFile
 }
 
+// Dialer interface to make getRemoteConnection testable.
+type dialerInterface interface {
+	Dial(string, string) (net.Conn, error)
+}
+
+type netDialer struct{}
+
+func (d *netDialer) Dial(network, address string) (net.Conn, error) {
+	return net.Dial(network, address)
+}
+
+func newNetDialer() dialerInterface {
+	return &netDialer{}
+}
+
+type tlsDialer struct{
+	config *tls.Config
+}
+
+func (d *tlsDialer) Dial(network, address string) (net.Conn, error) {
+	return tls.Dial(network, address, d.config)
+}
+
+func newTlsDialer(config *tls.Config) dialerInterface {
+	return &tlsDialer{config: config}
+}
+
 // GetRemoteConnection returns a connection to a remote address
 func GetRemoteConnection(useTLS bool, export *registry.ExportDetails) (remote net.Conn, err error) {
-	isLocalAddress := IsLocalAddress(export.PrivateIP)
+	var dialer dialerInterface
+	if useTLS {
+		config := tls.Config{InsecureSkipVerify: true}
+		dialer = newTlsDialer(&config)
+	} else {
+		dialer = newNetDialer()
+	}
+	return getRemoteConnection(export, dialer)
+}
 
+func getRemoteConnection(export *registry.ExportDetails, dialer dialerInterface) (remote net.Conn, err error) {
+	isLocalAddress := IsLocalAddress(export.HostIP)
+	
 	if isLocalAddress {
 		// if the address is local return the connection
 		address := fmt.Sprintf("%s:%d", export.PrivateIP, export.PortNumber)
-		return net.Dial("tcp4", address)
+		return dialer.Dial("tcp4", address)
 	}
 
 	// set up the remote address
 	remoteAddress := fmt.Sprintf("%s:%d", export.HostIP, export.MuxPort)
-	if useTLS {
-		config := tls.Config{InsecureSkipVerify: true}
-		remote, err = tls.Dial("tcp4", remoteAddress, &config)
-	} else {
-		remote, err = net.Dial("tcp4", remoteAddress)
+	remote, err = dialer.Dial("tcp4", remoteAddress)
+
+	// Prevent a panic if we couldn't connect to the mux.
+	if err != nil {
+		return nil, err
```

Comment:
If the mux connection fails, it returns to prevent remote.Write() below from causing a panic.  We have to write the mux header on remote connections though?  Maybe I'm misreading the comment though.